apps/heft/src/utilities/fileSystem/IExtendedFileSystem.ts

```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { FileSystem, IFileSystemCreateLinkOptions } from '@rushstack/node-core-library';
+
+export type FilesystemType = Omit<typeof FileSystem, 'prototype'>;
```

Comment:
There are two kinds of API interfaces:

- **easy-to-use**: A rich set of different operations for various usage scenarios, with lots of handy options
- **easy-to-implement**: A minimal set of "axioms" that fulfill the core functions needed to support higher-level systems

The design requirements are diametrically opposite for these two cases, so it's common to find them connected by an adapter that maps all the **easy-to-use** APIs onto a barebones plugin layer.

The `FileSystem` signature is designed to be **easy-to-use**, whereas your `CachedFileSystem` seems like it needs an `easy-to-implement` contract.  Otherwise every little change we make to `FileSystem` will break Heft and require special fixups to adapt it for caching (if I understand correctly).  Also it is telling that:

- `IExtendedFileSystem` has to add extra operations like `readFolderFilesAndDirectories()` that could not be cached reasonably
- Quite a lot of the `IExtendedFileSystem` are never actually used

For a good example of an **easy-to-implement** abstraction, take a look at the compiler's `ts.System` (which defines the bare minimum set of OS operations required to port the compiler to a new JavaScript runtime).  

Fixing this might be out of scope for this PR, but it seems like `CachedFileSystem` is going to run into architectural problems with its current design. And cache bugs are not fun bugs. :-)

Reply:
Yeah I don't like this design. We should redesign this, but let's do that in a separate PR.